app/src/main/java/org/dicekeys/app/encryption/BiometricsHelper.kt

```diff
@@ -16,87 +21,148 @@ import org.dicekeys.dicekey.Face
+            fragment.errorDialog(e)
         } catch (e: Exception) {
-            e.printStackTrace()
+            fragment.errorDialog(e)
         }
     }
 
-    private fun authenticate(biometricPrompt: BiometricPrompt, promptInfo: BiometricPrompt.PromptInfo, cryptoObject: BiometricPrompt.CryptoObject) {
-        biometricPrompt.authenticate(promptInfo, cryptoObject)
+    private fun authenticate(biometricPrompt: BiometricPrompt, promptInfo: BiometricPrompt.PromptInfo, cryptoObject: BiometricPrompt.CryptoObject?) {
+        if(cryptoObject == null){
+            biometricPrompt.authenticate(promptInfo)
+        }else{
+            biometricPrompt.authenticate(promptInfo, cryptoObject)
+        }
     }
 
-    private fun createBiometricPrompt(promptForEncryption: Boolean, fragment: Fragment): BiometricPrompt.PromptInfo {
+    private fun createBiometricPrompt(promptForEncryption: Boolean, keystoreType: AppKeystore.KeystoreType, fragment: Fragment): BiometricPrompt.PromptInfo {
         val title: String
         val subtitle: String
 
-        if(promptForEncryption){
+        if (promptForEncryption) {
             title = fragment.getString(R.string.biometrics_encryption_title)
-            subtitle = fragment.getString(R.string.biometrics_encryption_subtitle)
-        }else{
+            subtitle = if (keystoreType == AppKeystore.KeystoreType.BIOMETRIC) {
+                fragment.getString(R.string.biometrics_encryption_subtitle)
+            }else{
+                fragment.getString(R.string.authenticate_encryption_subtitle)
+            }
+        } else {
             title = fragment.getString(R.string.biometrics_decryption_title)
-            subtitle = fragment.getString(R.string.biometrics_decryption_subtitle)
+            subtitle = if (keystoreType == AppKeystore.KeystoreType.BIOMETRIC) {
+                fragment.getString(R.string.biometrics_decryption_subtitle)
+            }else{
+                fragment.getString(R.string.authenticate_decryption_subtitle)
+            }
         }
 
-        return BiometricPrompt.PromptInfo.Builder()
-                .setTitle(title)
-                .setSubtitle(subtitle)
-                .setNegativeButtonText(fragment.getString(android.R.string.cancel))
-                .setConfirmationRequired(true)
-                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
-                .build()
+        return BiometricPrompt.PromptInfo.Builder().let {
+            it.setTitle(title)
+                    .setSubtitle(subtitle)
+                    .setConfirmationRequired(true)
+
+            if (keystoreType == AppKeystore.KeystoreType.BIOMETRIC) {
+                it.setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
+                it.setNegativeButtonText(fragment.getString(android.R.string.cancel))
+            }else{
+                if(Build.VERSION.SDK_INT == Build.VERSION_CODES.R){
```

Comment:
What special about android R?

Reply:
From the documentation in `setAllowedAuthenticators`, some combinations are not available in older SDK.

```
             /**
             * Optional: Specifies the type(s) of authenticators that may be invoked by
             * {@link BiometricPrompt} to authenticate the user. Available authenticator types are
             * defined in {@link Authenticators} and can be combined via bitwise OR. Defaults to:
             * <ul>
             *     <li>{@link Authenticators#BIOMETRIC_WEAK} for non-crypto authentication, or</li>
             *     <li>{@link Authenticators#BIOMETRIC_STRONG} for crypto-based authentication.</li>
             * </ul>
             *
             * <p>Note that not all combinations of authenticator types are supported prior to
             * Android 11 (API 30). Specifically, {@code DEVICE_CREDENTIAL} alone is unsupported
             * prior to API 30, and {@code BIOMETRIC_STRONG | DEVICE_CREDENTIAL} is unsupported on
             * API 28-29. Setting an unsupported value on an affected Android version will result in
             * an error when calling {@link #build()}.
             *
             * <p>This method should be preferred over {@link #setDeviceCredentialAllowed(boolean)}
             * and overrides the latter if both are used. Using this method to enable device
             * credential authentication (with {@link Authenticators#DEVICE_CREDENTIAL}) will
             * replace the negative button on the prompt, making it an error to also call
             * {@link #setNegativeButtonText(CharSequence)}.
             *
             * <p>If this method is used and no authenticator of any of the specified types is
             * available at the time {@code authenticate()} is called,
             * {@link AuthenticationCallback#onAuthenticationError(int, CharSequence)} will be
             * invoked with an appropriate error code.
             *
             * @param allowedAuthenticators A bit field representing all valid authenticator types
             *                              that may be invoked by the prompt.
             * @return This builder.
             */
```